### PR TITLE
Add concurrency to the fetchData function

### PR DESCRIPTION
**Pull request by amdprophet**
Signed-off-by: Justin Kolberg <amd.prophet@gmail.com>

## Description
Adds concurrency to `fetchData()` in `daemon.go`. Sensu API performance becomes the main limiting factor in Uchiwa performance with this change.

## Related Issue
https://github.com/sensu/uchiwa/issues/750

## Motivation and Context
Uchiwa becomes very slow with one or more large Sensu installations.

## How Has This Been Tested?
Benchmarked against three staged Sensu datacenters:

Old:

```
{"timestamp":"2018-05-31T10:20:01.049587855-07:00","level":"warn","message":"Loading the configuration file config.json"}
{"timestamp":"2018-05-31T10:20:01.051740115-07:00","level":"info","message":"Updating the datacenter Site 1"}
{"timestamp":"2018-05-31T10:20:01.053135187-07:00","level":"warn","message":"Uchiwa is now listening on 0.0.0.0:3000"}
{"timestamp":"2018-05-31T10:20:16.863954073-07:00","level":"info","message":"Updating the datacenter Site 2"}
{"timestamp":"2018-05-31T10:20:36.863593601-07:00","level":"info","message":"Updating the datacenter Site 3"}
52.025680442s
{"timestamp":"2018-05-31T10:25:14.225580405-07:00","level":"info","message":"Updating the datacenter Site 1"}
{"timestamp":"2018-05-31T10:25:26.668433934-07:00","level":"info","message":"Updating the datacenter Site 2"}
{"timestamp":"2018-05-31T10:25:41.266359255-07:00","level":"info","message":"Updating the datacenter Site 3"}
38.911402966s
```

New:

```
{"timestamp":"2018-05-31T14:07:22.552640113-07:00","level":"warn","message":"Loading the configuration file config.json"}
{"timestamp":"2018-05-31T14:07:22.553484743-07:00","level":"warn","message":"Uchiwa is now listening on 0.0.0.0:3000"}
{"timestamp":"2018-05-31T14:07:22.553752527-07:00","level":"info","message":"Updating the datacenter Site 3"}
{"timestamp":"2018-05-31T14:07:22.554319529-07:00","level":"info","message":"Updating the datacenter Site 1"}
{"timestamp":"2018-05-31T14:07:22.554448697-07:00","level":"info","message":"Updating the datacenter Site 2"}
15.770854874s
{"timestamp":"2018-05-31T14:08:24.051077286-07:00","level":"info","message":"Updating the datacenter Site 3"}
{"timestamp":"2018-05-31T14:08:24.051162268-07:00","level":"info","message":"Updating the datacenter Site 1"}
{"timestamp":"2018-05-31T14:08:24.051213996-07:00","level":"info","message":"Updating the datacenter Site 2"}
14.344004765s
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: uchiwa/daemon/daemon.go [hunk 1]
```diff
@@ -66,124 +97,247 @@ func (d *Daemon) buildData() {
...
+	go d.fetchStashes()
+	go d.fetchSilenced()
+	go d.fetchChecks()
+	go d.fetchClients()
+	go d.fetchEvents()
+	go d.fetchInfo()
+	go d.fetchAggregates()
+
+	if f.enterprise {
+		go d.fetchEnterpriseMetrics()
```
Comment (echlebek) [inline uchiwa/daemon/daemon.go]: This probably requires a wg.Add(1)

File: uchiwa/daemon/daemon.go [hunk 2]
```diff
@@ -66,124 +97,247 @@ func (d *Daemon) buildData() {
...
+
+	d := DatacenterSnapshotFetcher{
+		snapshot:   &DatacenterSnapshot{},
+		datacenter: f.datacenter,
+		mutex:      mutex,
+		wg:         wg,
+	}
+
+	// fetch sensu data from the datacenter
+	wg.Add(8)
```
Comment (echlebek) [inline uchiwa/daemon/daemon.go]: Seems like this should be wg.Add(7)

File: uchiwa/daemon/daemon.go [hunk 3]
```diff
@@ -66,124 +97,247 @@ func (d *Daemon) buildData() {
...
+		d.snapshot.Stashes = append(d.snapshot.Stashes, v)
 	}
+
+	d.mutex.Unlock()
+	d.wg.Done()
+}
+
+func (d *DatacenterSnapshotFetcher) fetchSilenced() {
+	silenced, err := d.datacenter.GetSilenced()
+	d.mutex.Lock()
```
Comment (echlebek) [inline uchiwa/daemon/daemon.go]: A better pattern is
```
d.mutex.Lock()
defer d.mutex.Unlock()
```

This ensures that the mutex is always unlocked.

File: uchiwa/daemon/daemon.go [hunk 4]
```diff
@@ -66,124 +97,247 @@ func (d *Daemon) buildData() {
...
+			"datacenter %s. Silencing might not be possible, please update Sensu", d.datacenter.Name)
+	}
+
+	for _, v := range silenced {
+		setDc(v, d.datacenter.Name)
+		d.snapshot.Silenced = append(d.snapshot.Stashes, v)
+	}
+
+	d.mutex.Unlock()
+	d.wg.Done()
```
Comment (echlebek) [inline uchiwa/daemon/daemon.go]: `defer d.wg.Done()` at the top of the function.

File: uchiwa/daemon/daemon.go (hunk 2 above)
Comment (amdprophet) [reply uchiwa/daemon/daemon.go]: Fixed in the latest commit.

File: uchiwa/daemon/daemon.go (hunk 1 above)
Comment (amdprophet) [reply uchiwa/daemon/daemon.go]: Fixed in the latest commit.

File: uchiwa/daemon/daemon.go (hunk 4 above)
Comment (amdprophet) [reply uchiwa/daemon/daemon.go]: Changed - and also made use of the mutex in `fetchEvents()`.

**Pull request merged**